src/Parser/Import.ml

```diff
@@ -0,0 +1,177 @@
+(* This file is part of DBL, released under MIT license.
+ * See LICENSE for details.
+ *)
+
+(** Find and parse imported modules *)
+
+module StrSet = Set.Make(String)
+module StrMap = Map.Make(String)
+
+type import_set = StrSet.t
+
+let import_set_empty = StrSet.empty
+
+(** Produce the string representation of a relative path. *)
+let string_of_rel_path (p, n) =
+  List.fold_right (fun n' id -> n' ^ "/" ^ id) p n
+
+(** Produce the string representation of an absolute path. *)
+let string_of_abs_path p =
+  "/" ^ string_of_rel_path p
+
+(** Produce the string representation of a path. *)
+let string_of_path (p : Raw.import_path) =
+  match p with
+  | IPAbsolute(p, n) -> string_of_abs_path (p, n)
+  | IPRelative(p, n) -> string_of_rel_path (p, n)
+
+(** For an absolute module path produce a string suitable for use as an
+    internal module identifier. *)
+let internal_mod_id = string_of_abs_path
+
+(** Find the given file in one of the file system directories in [dirs]. *)
+let find_file_in dirs path =
+  dirs |> List.find_map (fun dirname ->
+    let fname = Filename.concat dirname path in
+    if Sys.file_exists fname then Some fname else None)
+
+(** Convert a module path to a (relative) file path to be found within the
+    file system. *)
+let to_file_path (p, n) =
+  List.fold_right Filename.concat p (n ^ Config.src_extension)
+
+(** Find the file system path for a given absolute module path. *)
+let find_mod_abs (p, n) =
+  match p with
+  | pref :: p when pref = Config.local_mod_prefix ->
+    find_file_in !Config.local_search_dirs (to_file_path (p, n))
+  | [] | _ :: _ ->
+    find_file_in !Config.lib_search_dirs (to_file_path (p, n))
+
+(** Find a module relative to the module directory path [prefix], and return
+    the absolute module path and file system path. *)
+let find_mod_rel prefix p =
+  let rec loop prefix_r (p, n) =
+    let abs_path = (List.rev_append prefix_r p, n) in
+    match find_mod_abs abs_path with
+    | Some fname -> Some (abs_path, fname)
+    | None       ->
+      begin match prefix_r with
+      | []            -> None
+      | _ :: prefix_r -> loop prefix_r (p, n)
+      end
+  in
+  loop (List.rev prefix) p
+
+(** Find the absolute module path and file system path for an import path. *)
+let find_mod prefix (p : Raw.import_path) =
+  match p with
+  | IPAbsolute(p, n) -> find_mod_abs (p, n) |> Option.map (fun f -> (p, n), f)
+  | IPRelative(p, n) -> find_mod_rel prefix (p, n)
+
+(** Find and parse imported module relative to the path [prefix] and add it
+    to the [mods] map with the absolute mod identifier as the key.
+    Previously imported modules can be passed in the [imported] parameter
+    to avoid including them twice. *)
+let rec parse_import ~imported prefix mods (import : Raw.import) =
```

Comment:
It took me a while before I realized what the [mods] map really is. It is a graph of dependencies. Maybe some comment, or locally defined type synonym would help.